### PR TITLE
update btcec to v2 and use the latest btcutil

### LINKED ISSUE #9249: Update github.com/btcsuite/btcd versions
## Summary

Currently, `tendermint/tendermint` is using `github.com/btcsuite/btcd v0.22.1`  which is a relatively old version of `btcsuite/btcd` and when it is used with some other modules with a more updated version of `btcsuite/btcd` like `v0.23.1`, the `btcsuite/btcd` being used for the whole project will be upgraded `v0.23.1` and it will cause some dependency conflicts due to the vast differences between `v0.23.1` and `v0.22.1`.

## Problem Definition

Currently, `tendermint/tendermint` can't be used with projects using newer version of `btcsuite/btcd` and of cause, `tendermint/tendermint` can't use newer version `btcsuite/btcd` either due to the dependency of old version(v0.22.1) of `btcsuite/btcd` in `crypto/scep256k1`. With this update, `tendermint/tendermint` could be more freely imported with other projects and `tendermint/tendermint` will also be able to use the latest version of `btcsuite/btcd`. There is basically no disadvantage to the tendermint.

## Proposal

Instead of using `github.com/btcsuite/btcd v0.22.1` and `github.com/btcsuite/btcutil v1.0.3-0.20201208143702-a53e38424cce`, we use `github.com/btcsuite/btcd/btcec/v2 v2.2.1` and `github.com/btcsuite/btcd/btcutil v1.1.2` which are the only two modules from `btcsuite` `tendermint` uses. It should eliminates all potential dependency conflicts with modules using the most updated `github.com/btcsuite/btcd`.

Some logic in `github.com/tendermint/tendermint/crypto/secp256k1` will have to be updated for the corresponding updates in dependencies.

### PR DESCRIPTION
**Pull request by wcsiu**
---

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [x] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

Closes #9249

### CODE REVIEW

**Comment by cmwaters** [comment]
We have recently changed the default branch to `main`. Could you rebase your changes of that branch?

Is `v2` backwards compatible. i.e. will v2 secp256k1 still verify old signatures?

**Comment by wcsiu** [comment]
> We have recently changed the default branch to `main`. Could you rebase your changes of that branch?
> 
> Is `v2` backwards compatible. i.e. will v2 secp256k1 still verify old signatures?

I have checked the test cases from both the older and newer version. The are the same.
https://github.com/btcsuite/btcd/blob/btcec/v2.2.1/btcec/ecdsa/signature_test.go#L40
https://github.com/btcsuite/btcd/blob/v0.22.1/btcec/signature_test.go#L38

I have also updated the tests in tendermint to make sure the signature verification tests pass.

Hence, the new version should still be able to verify old secp256k1 signatures.

Of course, as always, more tests on it will be better.

I will finish the rebase as soon as possible, but it is not trivial. In the worst case, I will submit a new PR.

**Comment by wcsiu** [comment]
@cmwaters I have rebased my branch to `main` branch. Please have a look.

**Comment by odeke-em** [comment]
Kindly cc-ing @veorq and @elias-orijtech to help with a security review.

**Review by elias-orijtech — COMMENTED** [review]
LGTM other than a suggestion about splitting the change in two; leaving for @veorq.

**Review by thanethomson — COMMENTED** [review]
Apart from the minor changelog nit, I'd recommend considering the change mentioned by @elias-orijtech.

**Comment by thanethomson** [comment]
@wcsiu just checking in here. This seems like a valuable change for us to have, but it needs some minor updates.

**Comment by wcsiu** [comment]
> @wcsiu just checking in here. This seems like a valuable change for us to have, but it needs some minor updates.

I agree with your recommendations. I will update the code according to them. I am having a busy week. I will try to finish it in the coming few days. Thank you.

**Comment by thanethomson** [comment]
> I agree with your suggestions. I will update the code according to them. I am having a busy week. I will try to finish it in the coming few days. Thank you.

No rush :slightly_smiling_face: Just wanted to check in. Thanks very much!

File: crypto/secp256k1/secp256k1.go [hunk 1]
```diff
@@ -122,24 +123,18 @@ func GenPrivKeySecp256k1(secret []byte) PrivKey {
...
-var secp256k1halfN = new(big.Int).Rsh(secp256k1.S256().N, 1)
-
 // Sign creates an ECDSA signature on curve Secp256k1, using SHA256 on the msg.
 // The returned signature will be of the form R || S (in lower-S form).
 func (privKey PrivKey) Sign(msg []byte) ([]byte, error) {
-	priv, _ := secp256k1.PrivKeyFromBytes(secp256k1.S256(), privKey)
+	priv, _ := secp256k1.PrivKeyFromBytes(privKey)
 
-	sig, err := priv.Sign(crypto.Sha256(msg))
+	sig, err := ecdsa.SignCompact(priv, crypto.Sha256(msg), false)
```
[the remark this answers is not in the document]
Comment (wcsiu) [reply crypto/secp256k1/secp256k1.go:131]: Just found that they no longer expose `R` and `S` as exported fields anymore(https://github.com/decred/dcrd/blob/dcrec/secp256k1/v4.1.0/dcrec/secp256k1/ecdsa/signature.go#L52) so logic in `serializeSig` will no longer works. 

Do you think we still need to break it into two commits?

[the remark this answers is not in the document]
Comment (elias-orijtech) [reply crypto/secp256k1/secp256k1.go:131]: I see, thanks. In that case, using `SignCompact` is probably ok.

[the remark this answers is not in the document]
Comment (wcsiu) [reply crypto/secp256k1/secp256k1.go:131]: OK.

**Comment by wcsiu** [comment]
@thanethomson @elias-orijtech has agreed the fix is no longer necessary. I think the PR should now be ready.

**Review by thanethomson — APPROVED** [review]

**Pull request merged**